stdlib/src/crypto/dsa/rpo_falcon512.rs

```diff
@@ -0,0 +1,149 @@
+use alloc::vec::Vec;
+
+use processor::{AdviceProvider, AdviceSource, ProcessState, handlers::EventError};
+use vm_core::{Felt, Word};
+
+/// Host handler function which pushes values onto the advice stack which are required for
+/// verification of a DSA in Miden VM.
+///
+/// Inputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: \[ SIGNATURE \]
+///
+/// Outputs:
+///   Operand stack: [PK, MSG, ...]
+///   Advice stack: [...]
+///
+/// Where:
+/// - PK is the digest of an expanded public.
+/// - MSG is the digest of the message to be signed.
+/// - SIGNATURE is the signature being verified.
+///
+/// The advice provider is expected to contain the private key associated to the public key PK.
+///
+/// # TODO:
+/// Eventually, we might want to make this a handler stateful, containing the signing keys.
+/// These could be loaded into the advice provider by the host when loaded into the Host.
+pub fn falcon_sig_to_stack_handler(
+    advice_provider: &mut dyn AdviceProvider,
+    process: ProcessState,
+) -> Result<(), EventError> {
+    let pub_key = process.get_stack_word(0);
+    let msg = process.get_stack_word(1);
+
+    let sk_felts = advice_provider
+        .get_mapped_values(&pub_key)
+        .ok_or_else(|| FalconError::NoSecretKey { key: pub_key }.into())?;
+
+    let result = falcon_sign(sk_felts, msg)
+        .ok_or_else(|| FalconError::MalformedSignatureKey { key_type: "RPO Falcon512" }.into())?;
+
+    for r in result {
+        advice_provider.push_stack(AdviceSource::Value(r)).expect("stack push failed");
+    }
+    Ok(())
+}
+
+// FALCON SIGNATURE
+// ================================================================================================
+
+/// Signs the provided message with the provided secret key and returns the resulting signature
+/// encoded in the format required by the rpo_faclcon512::verify procedure, or `None` if the secret
+/// key is malformed due to either incorrect length or failed decoding.
+///
+/// The values are the ones required for a Falcon signature verification inside the VM and they are:
+///
+/// 1. The challenge point, a tuple of elements representing an element in the quadratic extension
+///    field, at which we evaluate the polynomials in the subsequent three points to check the
+///    product relationship.
+/// 2. The expanded public key represented as the coefficients of a polynomial of degree < 512.
+/// 3. The signature represented as the coefficients of a polynomial of degree < 512.
+/// 4. The product of the above two polynomials in the ring of polynomials with coefficients in the
+///    Miden field.
+/// 5. The nonce represented as 8 field elements.
+#[cfg(feature = "std")]
+pub fn falcon_sign(sk_felt: &[Felt], msg: Word) -> Option<Vec<Felt>> {
```

Comment:
Should this take the secret key as `SecretKey`? The assumption is that the user will convert the key from felts into `SecretKey` if needed before calling this function.

Separately (or maybe alternatively), should we provide a public function to convert the Falcon signature into `Vec<Felt>` returned from this function?

Reply:
Agreed. This could be part of a broader `FalconHandler` implementation, which could store a set of signing keys.